mongomock/aggregate.py

```diff
@@ -605,6 +609,43 @@ def _handle_array_operator(self, operator, value):
 
             return None if None in parsed_list else list(itertools.chain.from_iterable(parsed_list))
 
+        if operator == '$map':
+            if not isinstance(value, dict):
+                raise OperationFailure('$map only supports an object as its argument')
+
+            # NOTE: while the two validations below could be achieved with
+            # one-liner set operations (e.g. set(value) - {'input', 'as',
+            # 'in'}), we prefer the iteration-based approaches in order to
+            # mimic MongoDB's behavior regarding the order of evaluation. For
+            # example, MongoDB complains about 'input' parameter missing before
+            # 'in'.
+            for k in ('input', 'in'):
+                if k not in value:
+                    raise OperationFailure("Missing '%s' parameter to $map" % k)
+
+            for k in value:
+                if k not in {'input', 'as', 'in'}:
+                    raise OperationFailure('Unrecognized parameter to $map: %s' % k)
+
+            input_array = self._parse_or_nothing(value['input'])
+
+            if input_array is None or input_array is NOTHING:
+                return None
+
+            if not isinstance(input_array, list):
```

Comment:
Let's allow tuples as well

Reply:
Got it. Done! Thanks.

Note: I think it would be interesting to have a way to "canonicalize" every input document so that a single expected type for each bson type. That `to_hashable` function implemented in #737 is something along those lines but with only "comparison" in mind.

Reply:
Sure thing. Feel free to refactor and send a proposal as a separate PR. It's OK to start small and use the sanitizer only for one or two operators.